### ISSUE TITLE
Performance on large HTML graphs.

### ISSUE THREAD

**Issue by lawrence-dol** [leeoniya/domvm#83]
Opening a new issue only to get this on the radar; I am not able to prioritize this for the next two weeks as the CEO is in town an I am demoing the proof-of-concept and will be in strategic meetings. However, I *will* definitely come back to this...

On another (only tangentially related) thread I said:

> I do have to say, for my mildly complex view of a few thousand nodes there is noticeable lag that, so far, I can only attribute to view generation and downstream. But I haven't looked into it deeply since function is a much higher priority to me.

It turns out that when I dumped and formatted the HTML my "few thousand" is closer to 8K. I have all the major details of my UI in place, now.

I am seeing a rendering time of about 500ms to render. And about 700 ms in paint. This is on Firefox 47 and may be directly related to that. Chrome seems quite spritely by comparison, though still not quite as snappy as I'd hoped. The total is about a second of latency.

I'll report back here as I make progress. The reformatted HTML is attached if you are curious. Please understand that this is still just a prototype.

[Screen-80x24.txt](https://github.com/leeoniya/domvm/files/321442/Screen-80x24.txt)

Also for the curious, the perf graph:

![renderperf](https://cloud.githubusercontent.com/assets/8662551/16167754/f580304e-34ad-11e6-851d-81e1a964c4fb.PNG)

**Comment by leeoniya** [comment]
That's a doozy for sure.

But before getting into the weeds, here is something you can try that might help. The thousands of empty cells with placeholders:

`<td class="cell A20"><div class="text">&nbsp;</div></td>`

could be converted to

`.cell {min-width: 5px}` &
`<td class="cell A20"></td>`

This will cut your tree size by 50%. I don't know if this is workable in your case, but I would certainly strive to do this since the wins would be huge. It looks like this interface would be one where a fixed-width font can be used. In this case you could also set your table layout to `fixed` in the css and gain some paint performance there.

After this, i would probably turn every row into a subview and create an algorithm that returns `false` from the the `render()` of any row that has 0-diff.

I used an older gist [1] to convert the html to one large domvm template and got around 330ms of scripting and 64ms in painting (Chrome).

[1] https://gist.github.com/leeoniya/d9bfcf12a88c8261f19b

**Comment by lawrence-dol** [comment]
They're good ideas. It is a fixed width font and I already am using fixed on the table. I have a note to convert text runs to a single div spanning multiple columns. 

At present the placeholders are for uniformity, but now it's working I am thinking they turn out to be unnecessary.

One problem I have in Chrome is that the table cells are a couple of pixels wider than the characters on their own, which makes spanning unworkable unless I can solve it. Doesn't happen on Firefox or Edge.

**Comment by lawrence-dol** [comment]
@leeoniya 

> The thousands of empty cells with placeholders ... could be converted to `<td>&nbsp</td>`.

Attempting to do that is proving fragile because of the ambiguity between:

    ["td","\u00A0"] // content value

and

    ["td", [ "\u00A0" ]] // content tag

largely because I have program-generated sub-values that are difficult to make aware of this difference -- they just want to return an array of contents.

Is there really no good way to make the graph uniform without this ambiguity?

I know that this is probably impossible, but I am being consistently struck with the fact that an array tuple is a poor data-model for an HTML tag, which is logically:

    { tag: "xx", atr: { ... }, con: [ ... ] }

It's the fact that this is modeled as:

    [ "xx", { ... }, [ ... ] ]

that causes the ambiguity with a tag and a simple text value. `[ ... ]` is now not clearly either an inner tag or an inner text value. That ambiguity is the sole reason I've wrapped the text values in a "div".

Am I missing something subtle?

**Comment by leeoniya** [comment]
> Am I missing something subtle?

Not really. I mean, the ambiguity is there and I cannot remove it without breaking all the things (tm). There are two good ways to avoid it and maintain explicit arrays: https://jsfiddle.net/ot0jh69s/1/. You can easily do the necessary pre-checks in your hyperscript wrapper to use either of these methods.

**Comment by lawrence-dol** [comment]
I guess it's going to have to be `new String`, since concatenating to the parent array makes an absolute mess of addressing in to the structure, something I have to do.

It's so much cleaner if the content of a cell is consistently `displayGrid[row][CON][col][CON][n]` rather that either that or `displayGrid[row][CON][col][CON+n]`. Which impacts everything, where to look for content, how to add and delete content, etc.

**Comment by leeoniya** [comment]
That makes sense, just make sure you only use `new String` on the the first child to disambiguate the array rather than for all strings everywhere since it is 88% slower than a string literal [1]. This may not matter much in your specific case, but as always: profile, profile, profile!

[1] https://esbench.com/bench/576974dadb965b9a00965b45 (see results -> run)

**Comment by lawrence-dol** [comment]
> make sure you only use new String on the the first child

Yes, of course.

**Comment by lawrence-dol** [comment]
@leeoniya : Would you be willing to consider any of the following special values to indicate "no tag, simple content"?

    [ null   , "some text" ]
    [ ""     , "some text" ]
    [ "!text", "some text" ]

**Comment by leeoniya** [comment]
how about i add support for tagging explicit child arrays?

it will be cheap to check and easy to add with no side-effects.

enumerable in for-in loops (if you don't use them, this is cleaner)

```js
var x = [1,2,3];

x._expl = true;

for (var k in x)
	console.log(k);

x.forEach(function(v, i) {
	console.log(i);
});
```

non-enumerable

```js
var y = [1,2,3];

Object.defineProperty(y, '_expl', { enumerable: false, value: true });

for (var k in y)
	console.log(k);

y.forEach(function(v, i) {
	console.log(i);
});
```

**Comment by lawrence-dol** [comment]
> how about i add support for tagging explicit child arrays?

That seems like it will be brittle in the case of duplicating existing fragments of the tree (which I am doing during post-construction manipulation for dynamic enhancements). Some that was an explicit child array will silently become not one when it's copied.

I far prefer the more obvious use of a special tag name for text. Since trying to use `[null]` right now fails with a type error, it seems like `null` is a safe and reasonably intuitive choice. What else could a "null tag name" mean, if not "no tag, thanks"?

**Comment by lawrence-dol** [comment]
(Completely BTW, I have duplicated the functionality of about 20K likes of Java in 1.6K lines of JavaScript, so far).

**Comment by tamasnet** [comment]
What about an option which subtly changes how the contents are processed. So for example:
```
["td", {_expl: true}, ["plain_text"]]
```
would produce:
```
<td>plain_text</td>
```
instead of:
```
<td><plain_text/></td>
```
and:
```
["td", {_expl: true}, [["span", "plain_text"]]] // note additional array
```
would give:
```
<td><span>plain_text</span></td>
```

Client code could constant the `{_expl: true}` to reduce extra trivial object creation.

This seems less magic/surprising than adding a marker prop to the child array, and looks like it should be comparable for performance impact.

**Comment by lawrence-dol** [comment]
That's a little better and would work in this specific case, though it would mean that every tag will now have properties, of at least that value.

To me, it more intuitive to express "the tag for this node is null" than "the content has explicit arrays for sub-tags" which in and of itself needs explaining in the context of why there's even an ambiguity. 

The `null` tag results in every child being a tuple [3] where two and/or three can be omitted. Which in turn make addressing 100% consistent. Even trying the `new String` approach I am having to add conditionals that say, "if the first element is a String", which is already starting to get ugly.

It's also completely backward compatible so there's no gain there. 

It seems like it should be just as efficient as any other proposed way of expressing a text node, which is, in the end, I think what I want to be able to do. To me what I am trying to indicate is a text node, not something about the children of the parent. I think that's the clincher -- if the parent expresses this fact there's a brittle relationship with it's children. If the node expresses this fact, it's self-contained.

This is (far?) more obvious and consistent (and succinct):

    ["td", [
        [null  , "plain text"]
        ["span", "more plain text"]
    ],["td", [
        ["hr"]
        [null  , "plain text"]
        ["span", "more plain text"]
    ]

than:

    ["td", {_expl: true}, [
        ["plain text"]
        ["span", "more plain_text"]
    ]],["td",
        ["hr"]
        ["plain text"]
        ["span", "more plain_text"]
    ]]

**Comment by leeoniya** [comment]
@tamasnet 

> Client code could constant the `{_expl: true}` to reduce extra trivial object creation.

Currently, `props` objects are absorbed into the vtree and mutated, so this won't work.

@lawrence-dol 

I definitely want to avoid in-band signaling within strings, so `"!text"` and other special characters are a non-starter, but `[null,...]` or `["",...]` may be workable.

> That seems like it will be brittle in the case of duplicating existing fragments of the tree (which I am doing during post-construction manipulation for dynamic enhancements). Some that was an explicit child array will silently become not one when it's copied.

Can you provide an example (code) of how this happens? I cannot imagine a situation where an explicit array would need to morph into a unary tag. Doesn't this defeat the purpose of having it be uniform?

My reasoning behind tagging an existing array is that i'm not trying to accommodate a manually-authored-template style but rather want to provide a path for children generators to return a uniform structure that doesnt need an offset of 1/`null`. If you're making a generator already, then tagging the array before return is not that grotesque.

**Comment by lawrence-dol** [comment]
> Can you provide an example (code) of how this happens?

Note: In my specific use case, I can control this, but...

I need to create the visual effect of a window panel on top of the main display. To do that I create a new `div` containing a `table` in the top-left cell with a higher `z-index`, spanning the rows and columns for the content, then move the cell contents into the new table sort of like this:

    wdwrow.push(["td", displayGrid[row][2][col][2] ]);

With the proposed change to the `td` attributes, I would have to "remember" to do:

    wdwrow.push(["td", {_expl: true },displayGrid[row][2][col][2] ]);

That's a brittle dependency because the parent is expressing something concrete about the formation of it's children. My hypertext wrapper could take care of this, but not everyone will be using one.

In other words, the correctness of the formation of a child node should not be dependent on a flag in its container... just as a general rule of thumb.

> not trying to accommodate a manually-authored-template style but rather want to provide a path for children generators to return a uniform structure

Right. And to do that what a generator wants is to simply return a uniform 3-tuple for every node. What if the child node generation was a separate concern from the container and/or child-array? Then neither tagging the parent node or the child-array is clean.

And even machine-generated structures will be inspected for debugging and/or reference, and walked and processed. In my case every node is a 3-tuple and what I am aiming to do is to drop the individual letter-per-cell and change them to a text-run-spanning-cells, which means that many cells will now not contain a text element at at.

**Comment by lawrence-dol** [comment]
@leeoniya 

> not trying to accommodate a manually-authored-template style but rather want to provide a path for children generators to return a uniform structure

Further on this, I would like both, because in the end even the generator is code I need to write, look at and understand, and I would like a node to look consistent regardless of where I am seeing it.

**Comment by leeoniya** [comment]
@lawrence-dol why not just generate your uniform tuple templates and have a recursive post-procrssing step in the top-level render() that tags an `._expl = true` flag to all child arrays before returning the final template? you would get both uniformity, no intermediate decision making and good perf.

let's leave `{_expl: true }` alone to reduce noise. it was a decent idea, but i'm thinking `[null,...]` will be the way to go as the signal...i'm actually pretty sure this already happens, is there a case when  this os not working for you?

**Comment by lawrence-dol** [comment]
Of the three options, I prefer this order:

  1. Explicit node formulation: `[ null, "plain text" ]`.
  2. Parent node flag `[ "td", {_expl: true }, [["plain text"]]]`.
  3. Array tag: `[ "td", [["plain text"]]._expl=true]`.

What if one wants HTML entities? There it seems that the explicit node is helpful:

    [ null, { _raw: true }, "plain&nbsp;text" }

?

**Comment by lawrence-dol** [comment]
> i'm actually pretty sure this already happens, is there a case when this os not working for you?

I'll try it again, but when I've had it in the past (accidentally and I think I tried it because it was, to me, the obvious thing), the result has been a type-error.

**Comment by leeoniya** [comment]
the whole point of an array tag would be so you dont have `["td", [["plain text"]]._expl=true]` but `["td", ["plain text"]._expl=true]`

right now `[null, {_raw: true}, "plain&nbsp;text"]` will not work. i think it would get flattened, the null then spliced out leaving you with a props as a child. this can be changed to work. `[null, "text"]` should work though.

**Comment by lawrence-dol** [comment]
> `[null, "text"]` should work though.

You are right; it does. I'm pretty sure it didn't at one point, though. But I could be mistaken.

`[null,null,"text"]` works too, which meets my current 3-tuple uniformity requirement. I am back in business. :-)

Thanks.

**Comment by lawrence-dol** [comment]
(side-bar

> the whole point of an array tag would be so you don't have [the first child as an array].

Then that's exactly what I don't want. I am specifically looking for uniformity in the structure so that the first node is not an exception, but doesn't have to be wrapped in a tag for that to be true. 

What happens to the array after I hand it over to DOMVM is not my concern.
/)

**Comment by leeoniya** [comment]
flagging an explicit child array gives you the uniformity because every element in the array can be presumed to be a node. and you can transplant literal strings knowing they will always ve treated as nodes regardless of if they are the first child or not.

while it's not turtles all the way down, for the purpose of node removal/replacement/transplanting it is.

`[null, null, "moo"]` will be much more expensive because it will need to not only allocate/GC but also to flatten and splice out the nulls during processing (some of this can be optimized if we decide a null-tag subarray is always a text node).

anyways, i'll work on adding both features and you can pick the option that provides the desired tradeoffs in each case.

**Comment by leeoniya** [comment]
@lawrence-dol 

So if we assume that all arrays with `[0] === null` are text nodes, we fail a test. Taking a trip down memory lane [1][2], we see why. The question is, how many conditions must we test that will be both performant and non-confusing?

Testing for a length == 2 or 3 and ensuring there are no sub-arrays? The rabbit hole actually goes deeper and gets more expensive since there can be getter functions to test for each array element. Only handling explicit text nodes inside of explicit child arrays (NO)?

[1] https://github.com/leeoniya/domvm/issues/50#issuecomment-206049356
[2] https://github.com/leeoniya/domvm/blob/1.x-dev/test/tests.js#L1643-L1662

**Comment by lawrence-dol** [comment]
@leeoniya 

Ahhh, the joy that is engineering complex, yet intuitive systems. The golden rule of intuition is context, context, context.

I guess the first point is that I did indeed take your advice and no longer need the first `null` to be considered a null-child. I suppose "null" tags are currently working due to flattening?

I think I would consider my original expectation to be infeasible due to the imprecision of using array tuples for nodes instead of objects. With that edge-case dropped, doesn't an array with `[0]==null` work?

Hmmm. Would a solution that *allows* object nodes without requiring them be lean enough to be acceptable and would it be unambiguous, I wonder?

    [ "td",[{ tag: null, con: "plain text" }]]

Then someone in my situation could use objects for all nodes:

    { tag:"td", con:[{ tag: null, con: "plain text" }] }

**Comment by tamasnet** [comment]
Along similar lines, what about making the content array one of the props?

```
["td", {_content: ["whatever text", ["span", "maybe some tags"]]}]
```

**Comment by leeoniya** [comment]
> I suppose "null" tags are currently working due to flattening?

They work because all non-tag arrays gets flattened, and nulls are always removed from everywhere.

> Hmmm. Would a solution that allows object nodes without requiring them be lean enough to be acceptable and would it be unambiguous, I wonder?

Perhaps, though my concern is that always testing if everything is an object may impact performance. If this were to happen, though, the signature would match the to-be-created vnode which can be absorbed: `{ tag:"td", body:[{ tag: null, body: "plain text" }] }`. But this bypasses the object pooling that was responsible for a significant perf improvement in 1.2.4+. In this scenario, you'd want domvm to give you a hyperscript function that takes params and can re-use objects from the internal pool.

First, https://github.com/leeoniya/domvm/commit/1c761eb07c23b0ca69255a0a25a16c2db6e9f21c painlessly implements explicit array tagging...which also keeps 100% of performance.

Second, https://github.com/leeoniya/domvm/tree/explicit-text-nodes uses `""` in [0] to infer that the node is text. This kind of works, but brings some other things out of the woodwork that must be tested for. For isomorphism to work, domvm merges adjacent text nodes, which requires testing (literal = cheap, else expensive). If a text node is an only child, there is no `createTextNode` called (it's slow), this only happens for interspersed text nodes where there is no choice. If you want complete uniformity WRT a tuple, and still be able to pass a `null` for `props`...which currently is not supported (`["", null, "text..."]`, this must be added. Basically there's some work to do here which I am not sure is worth the benefit.

The same problems with assumed text nodes exist regarding the fact that their length must be 1-3, their props object can really only be used for `_raw`, they cannot have any children that are also arrays. They would have to have a structure that is a subset of domvm templates, which makes them more fragile, and the entire system less performant (almost certainly). You guys can take a stab at handling these cases, but for me explicit sub-array tagging is sufficient. I think the point about a parent controlling how children are rendered isnt accurate. You're just telling the parent how to parse its body and tagging the body container with this hint, which should always be known at the time of body generation.

**Comment by leeoniya** [comment]
@tamasnet 

i would like to avoid stuffing things into props, since creating these objects is not free and iterating each of them and nulling the special props (which is done internally) becomes more expensive as more are added.

**Comment by lawrence-dol** [comment]
Using `[null,null,"plain text"]` is working fine for me.

80% to 90% of the time per host screen is in Gecko and Graphics, which suggests to me that the real low hanging fruit is in simplifying the HTML. 

I am currently at 600 ms per host screen (in Firefox) and I really need to be at 250 for acceptable performance and 100 or less for great performance. Chrome is much faster (easily fast enough). Edge is similar to Firefox, but fails on one of my screens in hydrate node with: 

SCRIPT5007: Unable to get property 'removed' of undefined or null reference
AppScript (176,79)

Same content works with both Firefox and Chrome.

**Comment by leeoniya** [comment]
you're at 600 ms after the `<td class="cell A20">&nbsp;</td>` simplification? is this with using the `[null, null, text]` signature for all text nodes? if you have any `did*` hooks, you could try setting `domvm.view.config({repaint: false})` which prevents a forced repaint before flushing the `did*` hook queue...but you will lose dom liveness guarantees, (which you may not need though, i dunno for your app).

unfortunately without anything to look at i cannot provide further guidance aside from generic advice to simplify your tree and do selective redraw, maybe occlusion culling, maybe ditch absolute uniformity.

> SCRIPT5007: Unable to get property 'removed' of undefined or null reference AppScript (176,79)

this is pretty useless to me since i have no testcase to debug, ff and chrome work fine and all written tests pass down to ie9. please open an separate issue for this when you have more info.

**Comment by lawrence-dol** [comment]
> you're at 600 ms after the `<td class="cell A20">&nbsp;</td>` simplification?

Yes. Surprisingly, there was no noticeable difference.

> is this with using the `[null, null, text]` signature for all text nodes?

Yes, though performance measurements in Firefox show the lion's share of the time being in the graphics and browser rendering, not in the JavaScript code. 

Conversely, Chrome and Edge show 99% of the time being in redraw and an unidentified "Anonymous" function. But Chrome hardly counts as that redraw time is between 40ms and 100ms. Edge spends 1.7 seconds in redraw; but it unaccountably takes 217ms just to set the focus on the active input field.

> if you have any `did*` hooks

None.

> this is pretty useless to me since i have no testcase to debug

Of course, I was just mentioning in passing why I can't get good numbers for Edge.

What you might be able to comment on is the fact that Edge shows every single table cell being individually removed and re-added on each host screen transition. This is a bit surprising because the whole display panel is keyed and that key changes for each screen.

**Comment by leeoniya** [comment]
@lawrence-dol 

> whole display panel is keyed

as in, every cell has a key? if that's the case, you should consider removing the keys if possible and have a function that can locate them by position. the key lookup is basically a `for` loop since re-building the hash on each redraw is also expensive.

Another thing you should definitely try is to verify that domvm is doing exactly what you expect it to do (as if you were writing the DOM code by hand). There are numerous ways you could mis-structure your template that could prevent domvm from doing the optimal set of DOM operations...and I think that your quest for textNode uniformity may be preventing some significant optimizations.

I've made a tool that i use in all my tests that instruments the DOM to count how many times various operations are invoked [1] during a mount/redraw. It can tell you if you (or the lib) is doing too much. I've put together a simple example [2] that shows how burying the text nodes inside arrays kills performance by forcing "createTextNode + insertBefore" rather than a simple "textContent", and how the redraw perf takes a 2-3x hit because of it.

Open the console and look at the DOM ops and timings. Note that the resulting DOM is identical in both cases.

[1] https://github.com/leeoniya/domvm/blob/1.x-dev/test/lib/dominstr.js
[2] https://jsfiddle.net/79mvnv6p/

**Comment by lawrence-dol** [comment]
@leeoniya 

> as in, every cell has a key

Oh, no. Just the `div` that contains the `table`, as a whole.

> I think that your quest for textNode uniformity may be preventing some significant optimizations

Fair enough, but it not uniformity for it's own sake; is for the sake of the code that has to manipulate the grid before it's handed off to DOMVM.

I *suppose* I could process out the cell arrays before I hand it off, but I think by then the damage is done. Still it could avoid the doubled up `insertBefore` calls (why are there two for every node?).

BTW, I've tried using `DOMInstr` following your fiddle and it just logs `Object { _time: 0 }` to the console. I wrapped the redraw call like you did:

    /**/instr.start();
    vm.redraw();
    /**/console.log(instr.end());

**Comment by leeoniya** [comment]
> BTW, I've tried using DOMInstr following your fiddle and it just logs Object { _time: 0 } to the console.

you have to either do `domvm.view.config({debounce: false});` to disable rAF and make the redraw synchronous, or do start() in a willRedraw hook and stop in a didRedraw hook.

**Comment by lawrence-dol** [comment]
Huh. That's pretty cool.

Firefox:

    _time: 113
    className: 1961, 
    createElement: 1961, 
    createTextNode: 1920, 
    insertBefore: 3881, 
    nodeValue: 2, 
    removeChild: 3880, 
    setAttribute: 21, 
    textContent: 5, 

Edge:

    _time: 736,
    className: 1961,
    createElement: 1961,
    createTextNode: 1920,
    insertBefore: 3881,
    nodeValue: 2,
    removeChild: 3078,
    setAttribute: 21,
    textContent: 5

Chrome:

    _time : 98
    className : 1961
    createElement : 1961
    createTextNode : 1920
    insertBefore : 3881
    nodeValue : 2
    removeChild : 3880
    setAttribute : 21
    textContent : 5

**Comment by leeoniya** [comment]
if your interface/data is mostly static, that's waaay too much stuff going on for a redraw call. usually you'd want to see a bunch of nodeValue/className maybe a few setAtteibute and textContent. if you open up the console on the dbmonster bench in domvm's repo and use "step" for 100% mutations, you'll get the idea.

**Comment by lawrence-dol** [comment]
No, that's not the case. Every new set of data from the host completely rewrites the UI. 

However, I don't understand why there are 3880 removeChild calls and about the same insertBefore calls when there are only 1920 cells (maybe one for the content and one for the td itself, for remove and insert?); I expected that the table element itself would be deleted wholesale in one hit and then rebuilt since the key of the panel containing the table changes.

I think I'll try creating plain text nodes next and normalizing them into an array of nodes when/if I need to append something to that cell.

**Comment by lawrence-dol** [comment]
@leeoniya 

Post-processing the tree to flatten text-nodes after I am done manipulating it, results in:

Firefox (same time):

    _time: 113
    className: 1960
    createElement: 1960
    createTextNode: 5
    insertBefore: 1965
    nodeValue: 2
    removeChild: 1962
    setAttribute: 21
    textContent: 1919


Edge (worse time by 100%):

    _time: 1388,
    className: 1960,
    createElement: 1960,
    createTextNode: 5,
    insertBefore: 1965,
    nodeValue: 2,
    removeChild: 1962,
    setAttribute: 21,
    textContent: 1919

Chrome (same time):

    _time : 66
    className : 1960
    createElement : 1960
    createTextNode : 5
    insertBefore : 1965
    nodeValue : 2
    removeChild : 1962
    setAttribute : 21
    textContent : 1919

Despite the number of operations dropping by half for removeChild, and half the insertBefore operations becoming textContent.

**Comment by lawrence-dol** [comment]
I did find a 250ms delay I had hacked in and forgotten to remove, which wasn't showing up in the performance numbers, but was contributing to the overall sluggish feel. That's gone now.

**Comment by leeoniya** [comment]
if we assume, the timing is inaccurate (perhaps it is, 113 seems wrong from your prior timeline estimates anyhow), can you feel any difference or see it in the timeline? just this change should have already yielded a massive perf difference.

why are there are high removeChild counts? it has to do with the possibility of hooks needing to be fired on any deep child nodes or subviews. we can probably explicitly indicate that there are no deep hooks and have the lib kill everything in one go.

removeChild : 1962 and createElement : 1960 makes me also think that this table is in fact largely static and a lot of the cells can be reused between screens. no?

**Comment by lawrence-dol** [comment]
> can you feel any difference or see it in the timeline? just this change should have already yielded a massive perf difference.

Yes to both. Edge is still awful, barely usable now instead of unusable.

> a lot of the cells can be reused between screens. no?

Yeah - I'd forgotten about that. There are only two display "resolutions" 80x24 (1920 cells) and 132x27 (3564 cells). The cell classes, attributes and contents change, but the structure only changes when we transition from one to the other.

I still have the task of joining text runs together, which will help further since only about 10% of the cells correspond to the start text. But that requires I figure out why Chrome puts an extra pixel or so on either side of a character rendered in a table cell as compared to that same character rendered in a text string.

**Comment by leeoniya** [comment]
> But that requires I figure out why Chrome puts an extra pixel or so on either side of a character rendered in a table cell as compared to that same character rendered in a text string.

try using normalize.css to get uniform behavior across browsers.

> Yeah - I'd forgotten about that. There are only two display "resolutions" 80x24 (1920 cells) and 132x27 (3564 cells). The cell classes, attributes and contents change, but the structure only changes when we transition from one to the other.

you should be able to reuse the full 1920 cells across both views which would eliminate 1920 removes, 1920 creates and 1920 inserts. currently view unmounting will destroy the dom & state but there are a couple ways to have the old dom recycled.

**Comment by leeoniya** [comment]
@lawrence-dol here's the best-case naive scenario for these two table sizes (without app-specific optimizations, isolated redraw, etc) [1]

I get ~30ms switching time in FF and Chrome and sub-10ms redraw(). Edge does a lot worse with ~300ms craziness. Maybe there are some MS-specific optims, or just reducing the removeChild count could be sufficient...there's definitely some DOM op that Edge is just very slow at.

[1] https://jsfiddle.net/sr4cptjv/4/

**Comment by lawrence-dol** [comment]
I can replicate your results, but I can't see why my DOM structure is not differenced and yours is; indeed many of the cells are identical and should not need to be replaced.

Is it because the key to the parent has changed? But I need that change to re-init the view so that the pre-computed stuff happens only when the host-data changes. Do I need to key the table within the div?

    <div class="display-panel">  // this is keyed (it's the view)
        <table class="grid C80"> // this is not
            ...
        </table>
    </div>

**Comment by lawrence-dol** [comment]
@leeoniya 

What's the principle(s) behind how DOMVM decides that a specific HTML node has changed? I seem to be able to get it to replace everything wholesale, or not at all, depending on whether I have a key (being the host data identity) on the VM root or not.

Strangely, it does always update the cells that have augmented contents added to them after the fact. But if a cell just has a text value, it's not being changed. See pics.

![screen1](https://cloud.githubusercontent.com/assets/8662551/16283298/223c6306-3881-11e6-8bde-fb0440d9c7d9.png)
![screen2](https://cloud.githubusercontent.com/assets/8662551/16283297/223c5910-3881-11e6-93de-1a0641ec85b7.png)

**Comment by lawrence-dol** [comment]
(* BTW, your `DOMInstr` is copyrighted. Do you have any objection with me lifting it verbatim into my JavaScript library for future use without a dependency on the DBMonster demo? *)

**Comment by leeoniya** [comment]
> Is it because the key to the parent has changed?

yes. if the view's key/handle changes, the view and its dom gets destroyed. this is done to ensure that any unmanaged DOM state is also destroyed (focused elements, partial values in fields, custom added attrs/style/classes, etc). The reason that in my example the DOM is reused is because the view does not change its key/model/handle, (but it also never re-inits because of this).

> What's the principle(s) behind how DOMVM decides that a specific HTML node has changed? 

The patching process is like this:

1. Build new tree, this is now source of truth.
2. Walk new tree and old tree in parallel.
	a. Check old tree in same parent/index for a compatible DOM node donor.
	b. When donor is found, transplant DOM node to into new tree and patch props based oldVnode <> newVnode diff
3. Walk old tree and clean up/remove any non-transplanted nodes, unmount any components that have disappeared.
4. Force repaint, fire did* hooks.

How donor is found is based on whether it's a view or tag, but are only searched within the same parent. Views are matched by the tuple [viewFn, handle] where handle defaults to the model that's passed into the view (arg 2) else by any explicitly provided key (arg 3), including `false`. Tags are matched on tag type and/or node _key. This glosses over a lot of intricacies, but basically if [viewFn, handle] cannot be found in the new vtree, it is unmounted and its DOM is cleaned up.

So, how can we indicate that 2 views can be considered DOM-similar even if their identity tuples don't match? Because this is what would be needed to retain/recycle the DOM in the face of old view disappearance. One option is to ignore the key and assume that if the viewFns match, (essentially contructors), then the DOM can be similar enough. I have this kind of on the todo [1] but it only solves the matching-contructor case. If the keys change and the viewFns change, then domvm would need help.

What I could do is provide some callback by which the user could indicate if the incoming vm (or view definition) can be treated as a "similar" to the current view, which would force the old view's DOM to be recycled by returning a donor match.

> (* BTW, your DOMInstr is copyrighted. Do you have any objection with me lifting it verbatim into my JavaScript library for future use without a dependency on the DBMonster demo? *)

It's MIT licensed. Though I would appreciate if you can keep the attribution there in some form, feel free to use and/modify/distribute as you see fit. It has no DBmonster dependency, the file is fully self-contained.

[1] https://github.com/leeoniya/domvm/blob/1.x-dev/src/view.js#L590

**Comment by lawrence-dol** [comment]
> if the view's key/handle changes, the view and its dom gets destroyed.

I appreciate that there's going to subtle details here I am unaware of, but it seems to me that reinit'ing the view is orthogonal to whether the DOM *structure* (vs. the *content*) is materially different. At least insofar as I understand the nature of the key.

My present understanding is that the key governs whether the model has been updated in such a way as to require that the view be reinitialized. In my case (and OMMV) the view-init pre-calcs a bunch of stuff needed by the view which does not need to happen on every redraw, greatly reducing the per-redraw work.

Perhaps this insight indicates that the concerns of view-model changes and DOM structure have been inadvertently conflated?

> I would appreciate if you can keep the attribution

Copyright and license comment retained. Thank you.

**Comment by lawrence-dol** [comment]
@leeoniya 

A puzzling thing (to me) that remains though. If I make the view unkeyed, I can detect the data change in the view render function and do the recalc's there, returning the same structure with different content in the cells (and some cells changed with respect to classes), but that gives me the results above in my pics.

Is that because only the `td` elements which themselves have changed are updated? That would make sense because the classes on a cell containing fields are different from one containing only a character.

**Comment by leeoniya** [comment]
> Perhaps this insight indicates that the concerns of view-model changes and DOM structure have been inadvertently conflated?

They're conflated for the purposes of ensuring that any DOM state that is added ad-hoc (rather than through domvm's templates) is destroyed when another view re-inits over the old DOM. Additionally, we need to evaluate how cheap it would be to reuse the DOM, so component similarity also needs to be taken into consideration when deciding whether different-viewFn views are cheaply compatible.

> If I make the view unkeyed, I can detect the data change in the view render function and do the recalc's there, returning the same structure with different content in the cells (and some cells changed with respect to classes)

Yes, you can definitely do this (and my sample does this), but it's a hack to detect model changes from within render. I would prefer to at least provide a `willUpdate` hook to the vms that would be invoked when the model gets swapped so you can put your re-init/state-reset logic into that.

But even with that, it's still kind of a hack, you should be able to swap another keyed view in and have the DOM reused either by viewFn-sameness or an explicit signal like `canRecycle(newView)`.

In your case, do both views use the same viewFn, just with different models/keys or different viewFns with different models/keys? Might be useful to have some form of OOP-like `instanceof` or `vm.kind` property to help with swappability determination.

I'm currently testing a change to the recycling algo that would do an pre-pass to mark removed views that can be DOM-reused based on viewFn sameness, but since the search needs to be exhaustive, the speed penalty may be too large.

> Is that because only the td elements which themselves have changed are updated? That would make sense because the classes on a cell containing a field are different from one containing only a character.

If the view stays in place, then any template returned by render should be patched properly regardless of the needed mutations. If this is not happening try to make a reproducible testcase, it could be a bug.

**Comment by lawrence-dol** [comment]
> In your case, do both views use the same viewFn, just with different models/keys or different viewFns with different models/keys?

It's the same view and structurally identical, except for the consideration of 80x24 vs 132x27 row/column count, with different contents. Literally the same view, but a new function since it's following the usual pattern of returning a closed-over anonymous function.

> by viewFn-sameness or an explicit signal like `canRecycle(newView)`

I would lean towards both options, though I am not sure how you'd detect "sameness" in a way that consistent and reliable. It seems like a `canRecycle` hook would be perfect in this case. What would that look like?

    return ["div.display-panel",{
        _hooks: {
            canRecycle          : function(vFn) { return true; }
            }
        onkeydown               : inpHandler.keyDownEvent,
        onclick                 : clkhdl,
        style                   : { zIndex: Z_DISPLAY_GRID },
        },[
        ["table.grid.C"+hstdta.dspCol(),displayGrid],
        ]];


> If this is not happening try to make a reproducible testcase, it could be a bug.

I'll work on that now, it's definitely not being patched to be fully consistent with the returned view. The trick will be winnowing it down to the smallest replication, since the obvious basic case is demonstrably working.

**Comment by leeoniya** [comment]
i've pushed a `willUpdate` hook to 1.x-dev. it will get invoked when the model changes on `false`-keyed views. this should allow you to keep your render clean in persistent views and still reset internal state.

```js
function A(vm, model) {
	vm.hook({
		willUpdate: function(vm, newModel) {
			model = newModel;
		}
	});

	return function() {
		return ["div", model.text];
	};
}
```

have a peek a the tests [1] to see how it works with both explicit `vm.update(newModel)` and implicit model swapping if the view is declarative: `[View, newModel, false]`.

[1] https://github.com/leeoniya/domvm/commit/51cded65d60a20463089bc7ef82b6e08da9bf1cd

**Comment by leeoniya** [comment]
> It seems like a `canRecycle` hook would be perfect in this case. What would that look like?

```js
var hooks = {
	canRecycle: function(viewFn, model, key) {
		// ...test similarity
		return true;
	}
}

// imperative
vm.hook(hooks);

// declarative
[myView, model, key, {hooks: hooks}]
```


There would be some more internal tweaking to do since currently will/didUnmount fires as part of the final cleanup pass, but we would need to fire it for recycling in this case as well. willUnmount -> canRecycle -> didUnmount

**Comment by lawrence-dol** [comment]
@leeoniya 

The `willUpdate` hook appears to be working, with the `false` key. And the problem with not updating nodes is not there.

If I *omit* the key, the `willUpdate` hook never fires which, I presume, is expected.

An screen-to-screen transition now looks like:

    Firefox:

    { _time: 97, createElement: 8, createTextNode: 4, removeChild: 5, insertBefore: 12, setAttribute: 16, textContent: 10, nodeValue: 500, className: 614, value: 1 }

    Chrome

    { _time : 41 className : 616 createElement : 10 createTextNode : 5 insertBefore : 15 nodeValue : 500 removeChild : 4 setAttribute : 21 textContent : 9 }    

    Edge

    {_time: 332, className: 616, createElement: 10, createTextNode: 5, insertBefore: 15, nodeValue: 500, removeChild: 4, setAttribute: 21, textContent: 9}

Which brings Edge just into the tolerable range. Hopefully they will improve significantly by the end of the year.

**Comment by lawrence-dol** [comment]
> The name may also be different to be more clear, maybe `canDonateTo`

From my (admittedly limited and outside-in) perspective, `canRecyle` or `canReuse` is much clearer. FWIW.

**Comment by leeoniya** [comment]
> If I omit the key, the willUpdate hook never fires which, I presume, is expected.

this is expected, because an omitted (or null) key causes the model itself to become the key/handle (domvm opts for assumptions of model persistence, and you have to opt out of this if you have model swapping without implying view destruction).

For `update` to be allowed, the `key` must be set to something that is persistent across redraws, so..false, some fixed value (string/number), false, but the model itself.

**Comment by lawrence-dol** [comment]
> something that is persistent across redraws

Is there a conflict if two sub-views in the same parent have the same key?

**Comment by leeoniya** [comment]
only if they also have the same viewFn [1],

but in general, they're called *keys* for a reason ;)

[1] https://github.com/leeoniya/domvm/blob/1.x-dev/src/utils.js#L145

**Comment by lawrence-dol** [comment]
By "viewFn" you mean the view-constructor, right, not the view-renderer that it returns?

    function MyView(vm) {   // <= constructor
        // ...
        return function() { // <= renderer
            return ["h1","Hello"];
        }
    }

**Comment by leeoniya** [comment]
right

**Comment by lawrence-dol** [comment]
Please let me know (here) when the `canReuse` (or whatever) feature is available, and I'll switch to it. But my problem is resolved.

**Comment by leeoniya** [comment]
@lawrence-dol 

> An screen-to-screen transition now looks like...

ah, that's more like it :) plus you can optimize this further by creating self-diffing subviews for each row. which would shave another 10-20%. i'll ping you when/if i can get `canReuse` implemented with negligible reconciliation overhead.

plz keep this discussion in mind for writing docs, as the model/key/handle concept [and how it affects view recycling/destruction] is probably the most thorny topic to understand in the lib.

**Comment by lawrence-dol** [comment]
I can't edit other peoples issues, but you should add a reference in #83 to this.

**Comment by leeoniya** [comment]
btw, DOMInstr could probably be replaced by MutationObservers in newer browsers. Though i'm not sure if it would be possible to get the needed level of detail out of them.

[1] https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
